Mathlib/RingTheory/Ideal/Over.lean

```diff
@@ -429,4 +429,94 @@ lemma map_eq_top_iff {R S} [CommRing R] [CommRing S]
 
 end IsDomain
 
+section ideal_liesOver
+
+variable (A : Type*) [CommSemiring A] {B : Type*} [Semiring B] [Algebra A B]
+  (P : Ideal B) (p : Ideal A)
+
+/-- The ideal obtained by pulling back the ideal `P` from `B` to `A`. -/
+abbrev under : Ideal A := Ideal.comap (algebraMap A B) P
+
+theorem under_def : under A P = Ideal.comap (algebraMap A B) P := rfl
+
+instance IsPrime.under [hP : P.IsPrime] : (under A P).IsPrime :=
+  hP.comap (algebraMap A B)
+
+/-- If `P` is a maximal ideal of `B`, then the intersection of `P` and `A` is also
+  a maximal ideal. -/
+instance IsMaximal.under (A : Type*) [CommRing A] {B : Type*} [CommRing B] (P : Ideal B)
+    [P.IsMaximal] [Algebra A B] [Algebra.IsIntegral A B] : (Ideal.under A P).IsMaximal :=
+  isMaximal_comap_of_isIntegral_of_isMaximal P
+
+variable {A}
+
+/-- `P` lies over `p` if `p` is the preimage of `P` of the `algebraMap`. -/
+class LiesOver : Prop where
+  over : p = Ideal.comap (algebraMap A B) P
+
+instance over_under : P.LiesOver (Ideal.under A P) where over := rfl
+
+variable {A : Type*} [CommSemiring A] (p : Ideal A) {B : Type*} [CommSemiring B] (P : Ideal B)
```

Comment:
I'm not a fan of shadowing variables. Is it possible to split them into sections?
That is,
```lean
section Semiring
variable {A} [Semiring A]
-- stuff
end Semiring

section Ring
variable {A} [Ring A]
-- stuff
end Ring
```

Reply:
Also the order of the arguments of the ideals is different from the def. Maybe reverse it here?